### ISSUE TITLE
[rush] Simplify "rush deploy" configuration

### ISSUE THREAD

**Issue by devin-fee-ah** [microsoft/rushstack#1906]
**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**
Marked it as a bug, because I think other end-users will see this as a bug.

Rather than give the narrative, let me display the docs and ask the question:

>   /**
>      * Specify how links (symbolic links, hard links, and/or NTFS junctions) will be created in the deployed folder:
>      *
>      * - "default": Create the links while copying the files; this is the default behavior.
>      * - "script": A Node.js script called "create-links.js" will be written.  When executed, this script will create the links described in the "deploy-metadata.json" output file.
>      * - "none": Do nothing; some other tool may create the links later.
>    */

So what is the `default` behavior? (Hint, it's not `default`!)!

... it's `none`!

This has to change.

Proposal: how about:
`none` becomes `default`... and becomes the `undefined` option.
`default` becomes `create`
`script` remains `script`

@octogonz

**Comment by devin-fee-ah** [comment]
Thinking through it further, it seems the "lowest common denominator" pattern for a deployable is the same tree structure that a non-monorepo looks like:

```
- /
- /package.json
- /node_modules/**
- /**
```

Unfortunately, there's not a strategy to support this pattern, though `default` / `script` get somewhat close (the symlink strategy for non-windows still makes it incompatible).

Perhaps there's an opportunity to unbundle `linkCreation`:

```ts
{
  installation: {
    // how do you want the files to copied (currently hard-coded as `os === 'win' ? 'hardlink' : 'softlink'`
    method: 'symlink' | 'hardlink' | 'copy';
    // should the installation be performed now?
    strategy: 'defer' | 'execute'; 
  },
}```

While continuing to maintain the `default`

**Comment by devin-fee-ah** [comment]
As more time passes from my initial comment, it seems to me that there should probably be one `deployment.json` file that enumerates the various packages (that could be deployed) and their deployment options. (or better yet, since this is 1:1, maybe this actually belongs in `rush.json` itself). For everything else, it seems that `rush-lib` is the appropriate escape hatch.

The solution, as it stands seems to be a `pnpm lite`. Symlinks everywhere, but why? This is supposed to be our one-way exit from `pnpm` (or any other package manager) to lowest-common-denominator deployment.

**Comment by octogonz** [comment]
> > /**
> > * Specify how links (symbolic links, hard links, and/or NTFS junctions) will be created in the deployed folder:
> > *
> > * - "default": Create the links while copying the files; this is the default behavior.
> > * - "script": A Node.js script called "create-links.js" will be written.  When executed, this script will create the links described in the "deploy-metadata.json" output file.
> > * - "none": Do nothing; some other tool may create the links later.
> > */
> 
> So what is the `default` behavior? (Hint, it's not `default`!)!
> 
> ... it's `none`!
> 
> This has to change.

This is a bug. I was able to repro it.  The  default behavior is definitely supposed to be `default`. @devin-fee-ah Thank you for reporting!

**Comment by octogonz** [comment]
> Thinking through it further, it seems the "lowest common denominator" pattern for a deployable is the same tree structure that a non-monorepo looks like:
> 
> ```
> - /
> - /package.json
> - /node_modules/**
> - /**
> ```
> 
> Unfortunately, there's not a strategy to support this pattern, though `default` / `script` get somewhat close (the symlink strategy for non-windows still makes it incompatible).

We [mentioned this](https://github.com/microsoft/rushstack/issues/1078#issuecomment-636147289) in the design discussion:
> > (@iclanton has also suggested that `rush deploy` could flatten the monorepo folder structure, i.e. eliminating the `apps` and `libraries` levels. This is not too difficult to implement, but outside the scope of the MVP.)

I would like to implement a feature where the "main project" ends up in `common/deploy` and the other projects go in subfolders. However this is a nontrivial mapping, because it needs to preserve the PNPM installation plan. The MVP algorithm for crawling a subset of dependencies was already fairly challenging without this remapping.

> Perhaps there's an opportunity to unbundle `linkCreation`:
> 
> ```ts
> {
>   installation: {
>     // how do you want the files to copied (currently hard-coded as `os === 'win' ? 'hardlink' : 'softlink'`
>     method: 'symlink' | 'hardlink' | 'copy';
>     // should the installation be performed now?
>     strategy: 'defer' | 'execute'; 
>   },
> }```

We cannot really provide this choice between `'symlink' | 'hardlink' | 'copy'`.  

The `symlink`/`hardlink` distinction has really only one solution as far as I call (see all the if/then cases for [_deploySymlink()](https://github.com/microsoft/rushstack/blob/93b598f2013358e9c13c90f3da3ad11cefc1db79/apps/rush-lib/src/logic/deploy/DeployManager.ts#L417).

And regarding a `'copy'` option:

> Symlinks everywhere, but why? This is supposed to be our one-way exit from `pnpm` (or any other package manager) to lowest-common-denominator deployment.

The `rush deploy` mechanism reuses PNPM's installation plan, which inherently requires symlinks.  There are 3 popular ways to represent the "directed-acyclic-graph" of package dependencies:

- **NPM / ClassicYarn:** force the graph into a tree shape by introducing inaccuracies (phantoms/doppelgangers) that cause problems
- **PNPM:** use symlinks
- **Yarn Plug'n'Play:** redesign the Node.js `require()` API to understand **package.json** rather than traversing a `node_modules` hierarchy

These 3 installation plans are fundamentally different and there aren't conversions between them. (This is why there are 3 totally different lockfile formats.)

In short, `rush deploy` can provide various options for how the symlinks get created, but we can't eliminate them entirely.

**Comment by octogonz** [comment]
@devin-fee-ah This feedback was very valuable. Let me know if you have any further questions/thoughts.

**Comment by octogonz** [comment]
> This is a bug. I was able to repro it. The default behavior is definitely supposed to be `default`. @devin-fee-ah Thank you for reporting!

We are publishing a fix for this with PR https://github.com/microsoft/rushstack/pull/1907 , as Rush `5.24.1`.  Should be ready in a few minutes.

**Comment by dfee** [comment]
For the simple use case (e.g. generating an express app) "deployment" requires an extra degree of indirection – a "scenario".

Problems:
- (assuming this is going to follow the power-law, like everything else), the number of packages which will require [0, 1] scenarios is ~99%, and the number of package which will have 2+ scenarios is < 1%.
- `rush` shouldn't automate complex deployment processes, but instead accommodate them – with an API.
- mapping package names -> scenario names (for calls into `rush deploy`) ->  package names (so we can use `rushConfiguration.getProjectByName()` to get it's `common/deploy/<path>`) allows us to solve problems... that should probably be beyond the scope of this tool. (but not beyond the scope of custom tools that call into `rushConfiguration`!).
- the argument that multiple teams manage different deployments seems to be irrelevant. this problem is baked into the singular `rush.json`, `common-versions.json`, and other files. it doesn't feel like this is the right time to deviate from that design pattern.

(the quote):
> Lastly @iclanton told me he could easily imagine 30 different deployments, each managed by a different group of people who shouldn't have to approve each other's PRs. (That's why the scenarios go in separate files.)

I'd love for the following:
- `common/deploy` or `-t` is the actual target for the "deployment"... i.e hoist the target package to the root (this is on @octogonz 's roadmap, but punted for POC / MVP).
- one mapping of packageName -> deployment strategy (with discussion of advanced build patterns in documentation).

Etc.
- `deploy` doesn't seem *too different* from `publish`. Therefore, I think they could both be in `rush.json`.

**Comment by octogonz** [comment]
> `rush` shouldn't automate complex deployment processes, but instead accommodate them – with an API.

I prefer to avoid custom code wherever possible. It is "more flexible," but when you have to fix a bug in someone else's monorepo, life is much easier with:

- **standardized ways of doing things** so you know where to look to answer basic questions like "What gets deployed by this repo?"
- **config files instead of code** so you can answer those questions without having to figure out what a custom script does
- **ability to run it yourself** -- I want to do `rush deploy` on my laptop and see the same artifacts that the CI machine would produce.

Custom code works against all these benefits.

> the argument that multiple teams manage different deployments seems to be irrelevant. this problem is baked into the singular `rush.json`, `common-versions.json`, and other files. it doesn't feel like this is the right time to deviate from that design pattern.

🤔 This is a fair point.

> `deploy` doesn't seem _too different_ from `publish`. Therefore, I think they could both be in `rush.json`.

We considered this, but I think I'd prefer to avoid mixing together `rush publish` and `rush deploy` designs. Publishing involves NPM versioning, which is already a sufficiently complex problem.  It keeps getting procrastinated because the current model works tolerably, but at some point we're going to overhaul `rush publish`, and there is a laundry list of extra requirements it will bring (see https://github.com/microsoft/rushstack/issues/904 for some details). 

> For the simple use case (e.g. generating an express app) "deployment" requires an extra degree of indirection – a "scenario".

What if we made this change: **Scenarios are optional because there is a "default" one.**  By default, `rush deploy-init` will create `common/config/rush/deploy.json`, and that is what `rush deploy` will process. But of you do want multiple named scenarios, you an still create them under `common/config/deploy-scenarios`.

Would that address your concerns?  Just to be clear, `rush deploy` is marked as "experimental" for now. Breaking changes are fine if they make things better somehow.

**Comment by octogonz** [comment]
>     * one mapping of packageName -> deployment strategy (with discussion of advanced build patterns in documentation).

How would this work with subdeployments?

**Comment by dfee** [comment]
> Custom code works against all these benefits.

There is a limit here, of course. But, this isn't my specific gripe. It just seemed to encompass all my concerns. Let's move on.

> We considered this, but I think I'd prefer to avoid mixing together `rush publish` and `rush deploy` designs.

Again, fine. Good background on the problem landscape, though.

> What if we made this change: **Scenarios are optional because there is a "default" one**. By default, `rush deploy-init` will create `common/config/rush/deploy.json`, and that is what `rush deploy` will process. But if you do want multiple named scenarios, you an still create them under `common/config/deploy-scenarios`.

Now we're getting somewhere...

- If there's a default scenario – `deploy.json` – then it'll need to be parameterized based on the package name to be built. i.e. today `rush init-deploy` (😉) takes a parameter `--project`
- it will also need to accomodate (sane) default options or be parameterized itself: `includeDevDependencies`, `includeNpmIgnoreFiles`, and `linkCreation`.

> > one mapping of packageName -> deployment strategy (with discussion of advanced build patterns in documentation).

> How would this work with subdeployments?

I don't understand the value of `subdeployments` right now. This goes back to the very first point. I feel like `rush deploy` is pretty cheap. My ideal situation would be to iter over the projects to build. Just like you're going to have to already do for `rush build -t {<foo>,<bar>,<baz>}`.

**Comment by dfee** [comment]
Also, assuming that `rush init-deploy` generates the default `deploy.json`... which is parameterizable... then `deploy.json` should not exist IMO. there's no reason for the boilerplate `rush init-deploy` call.

**Comment by dfee** [comment]
Here's a quick-and-dirty example of what I'm thinking...

https://gist.github.com/dfee/13b7b5fa1197519699333c396629b824

**Comment by octogonz** [comment]
I just chatted with @iclanton about your suggestions. We also went back and re-read the requirements from @SoyYoRafa ( https://github.com/microsoft/rushstack/issues/1078#issuecomment-635681263 ) and @buffcode  ( https://github.com/microsoft/rushstack/issues/1078#issuecomment-635785568 ) that originally motivated the `enableSubdeployments` design.

Problems we're trying to solve with this simplification:

- Many users will deploy only one project. It seems redundant for them to have to specify a "scenario name" in addition to the deployment project name.
- When users deploy multiple projects, in most cases they will still share a single config file.  (But there are advanced cases where multiple config files are needed.)
- We want to keep the CLI simple. (For example, a big criticism of `rush publish` was that all those CLI parameters are confusing compared to a config file with nice doc comments.)

Also, @dfee pointed out that when multiple deployments happen, it is usually orchestrated by some CI script that needs to loop over all the projects (or is simply invoked multiple times on different CI machines).  Such a script can easily invoke `rush deploy` multiple times and specify target folders. Thus having `rush deploy` loop over "subdeployments" and create subfolders maybe is not the best design.

Here's what we came up with:

- The `--scenario` parameter will become optional.  The `rush init-deploy` and `rush deploy` commands will by default use `common/config/rush/deploy.json`.
- Whereas if you specify `--scenario foo`, the corresponding config file will be `common/config/rush/deploy-foo.json`. We are betting that there won't be enough of these files to justify tracking them in a special folder.
- Eliminate the `enableSubdeployments` feature.  `rush deploy` will always perform exactly one deployment.  No more looping over subdeployments in one command.
- Convert the `deploymentProjectNames` setting to be a list of projects that are *intended to be performed* by a given scenario file. We could remove `deploymentProjectNames` entirely, but @iclanton and I both felt strongly that an explicit list is important (1) to catch mistakes and (2) to document the intended workflow for your monorepo.
- For `rush deploy`, the `--project` parameter will become optional. It is only needed when `deploymentProjectNames` contains more than one item.

Sample setup:

**common/config/rush/deploy.json**
```js
{
  "deploymentProjectNames": [ "app1", "app2" ],
  "includeDevDependencies": false,
  "includeNpmIgnoreFiles": false,
  "linkCreation": "default",
  "projectSettings": [
    . . .
  ]
}
```

**common/config/rush/deploy-app2-aws.json**
```
{
  "deploymentProjectNames": [ "app2" ],
  "includeDevDependencies": true,
  "includeNpmIgnoreFiles": false,
  "linkCreation": "script",
  "projectSettings": [
    . . .
  ]
}
```

Sample CLI commands:
- `rush init-deploy --project app1` <br/> create the **deploy.json** file with `app1`.  (`app2` can be added manually.)
- `rush init-deploy --project app2 --scenario app2-aws` <br/> create the **deploy-app2-aws.json** file with `app2`
- `rush deploy --project app1` <br/> deploy `app1` from **deploy.json**.  Note that we need `--project` here because there are two `deploymentProjectNames` items. The output goes to **common/deploy** (not **common/deploy/app1**).
- `rush deploy --scenario app2-aws --target-folder /mnt/share` <br/> deploy `app2` from **deploy-app2-aws.json** The output goes to the **/mnt/share** folder.

Since the `rush deploy` feature is still "experimental," this redesign will get published as a minor version increment for Rush.

**Comment by octogonz** [comment]
Here's a PR: https://github.com/microsoft/rushstack/pull/1931

**Comment by octogonz** [comment]
This was published with Rush 5.26.0

The docs are here: https://rushjs.io/pages/maintainer/deploying/
